crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs

```diff
@@ -3290,9 +3293,27 @@ impl Analyzer<'_, '_> {
         }
 
         match new_ty.normalize() {
-            Type::Keyword(..) | Type::Lit(..) => {}
+            Type::Keyword(keyword) => {
+                if let TsKeywordTypeKind::TsObjectKeyword = keyword.kind {
+                    if orig_ty.is_any() {
+                        return Ok(orig_ty.into_owned());
+                    }
+                }
+            }
+
+            Type::Lit(..) => {}
             _ => {
                 match orig_ty.normalize() {
+                    Type::Keyword(KeywordType {
+                        kind: TsKeywordTypeKind::TsAnyKeyword,
+                        ..
+                    }) => {
+                        if new_ty.is_fn_type() {
+                            return Ok(orig_ty.into_owned());
+                        }
+
+                        return Ok(new_ty.into_owned());
+                    }
```

Comment:
This logic is based on this comment [// 'any' is narrowed to types other than 'Function'/'Object' (including {})](https://github.com/microsoft/TypeScript/blob/d87d0adcd30ac285393bf3bfbbb4d94d50c4f3c9/tests/baselines/reference/narrowFromAnyWithTypePredicate.errors.txt#L29)